### DISCUSSION TITLE
How to use HTTP3 with .NET 6 on Win7, Win10 and MacOS?

### DISCUSSION THREAD

**Discussion by alexandrehtrb** [microsoft/msquic#2151] [Q&A] [answered]
Hello!

I am very interested in HTTP3 and I would like to start using it.
According to the [Platforms](https://github.com/microsoft/msquic/blob/main/docs/Platforms.md) document, MsQuic should work on any OS that can run OpenSSL. If I want to use HTTP3 with .NET 6 on Win7, Win10 or MacOS, what do I need to do? I was thinking about this:

1) Have OpenSSL installed
2) Build a custom [System.Net.MsQuic.Transport](https://github.com/dotnet/msquic) lib, specifying `-Tls OpenSSL`
3) Reference this custom lib to my projects, it will automatically override the one from dotnet runtime

Is that it? Has anyone did something similar?

**Comment by nibanks** [comment, marked as answer]
cc @wfurt, @ManickaP

I believe .NET on Windows doesn't support OpenSSL as is, though it has been asked for in the past. The .NET devs would know more about how hard it would be to do this.

**Reply by ManickaP** [reply]
HTTP/3 is supported in .NET 6.0, on Windows it'll work out of the box, on Linux you need to get msquic first. Everything is covered in this blog post: https://devblogs.microsoft.com/dotnet/http-3-support-in-dotnet-6/#prerequisites, including necessary project settings and code example.

Is this enough info to get you going? Do you need more help with anything particular?

And please feel free to open issue in [runtime](https://github.com/dotnet/runtime) repo if you encounter any problems with HTTP/3 in .NET.

**Reply by nibanks** [reply]
@manickap I believe the ask here is to be able to use the OpenSSL variant of MsQuic on Windows, which isn't supported by .NET right now. The current set up will only work out of the box on Win11 or WS2022.

**Reply by ManickaP** [reply]
@alexandrehtrb are you interested in HTTP/3 in general or in QUIC with OpenSSL specifically?

@nibanks yes, the Windows version must be high enough. It's covered in the blog post I linked. I didn't want to copy paste everything here. There are other steps as well (like enabling preview features in .NET etc.)

**Reply by alexandrehtrb** [reply]
I am interested in HTTP3 in general, but by running .NET 6 on Win7, Win10 and MacOS.
From what I read [here](https://github.com/microsoft/msquic/blob/main/docs/Platforms.md), the only way to make .NET 6 run HTTP3 on those platforms would be by creating a custom [System.Net.MsQuic.Transport](https://github.com/dotnet/msquic) lib that uses OpenSSL. Is there an easier way?

**Reply by thhous-msft** [reply]
Even at that point, the Windows .NET 6 support is hardcoded to schannel, and will crash if an openssl msquic binary is used. And on macOS there is a known bug that causes macOS to not work in .NET 6. Getting things to work on those platforms will require a custom dotnet build to fix the issues, not just the correctly build msquic transport library (The mac bug is fixed in main already, but openssl on windows support I don't think is being worked on).

**Reply by alexandrehtrb** [reply]
I understand. Are there plans for supporting or installing Schannel on older Windows versions? The OS version restriction will leave some users left behind, it would be good to support older Windows versions.

**Reply by karelz** [reply]
To my best knowledge, Schannel does not plan to backport TLS 1.3 (required for QUIC and therefore HTTP/3) to older versions of Windows.
.NET does not have plans to support OpenSSL backend on Windows -- it is too much work for very little gain + suboptimal experience without OS integration.

Each product has to decide where they draw the line of backporting vs. adding features into future releases. Backporting features means increased risk of regressions, more work and should be therefore justified by strong user demand.
Personally, I don't think users will desperately need HTTP/3 and QUIC support anytime soon -- older HTTP versions are not going away and the people who care about the perf benefits of HTTP/3 will likely be willing to upgrade to newer OS version to receive even more perf benefits in other components.

**Comment by wfurt** [comment]
Is there tracking for the macOS issue @thhous-msft? I'm busy at the moment but improving macOS support is on my TODO list. We plan to include macOS support in .NET 7.0. 
Also since the question is .NET specific https://github.com/dotnet/runtime may be better place

**Reply by thhous-msft** [reply]
https://github.com/dotnet/runtime/issues/58849 It was this bug. Its been fixed, its just not in .NET 6
